Support fetching data with Request object

ShawnMa16 opened the pull request.

MarcoEidinger commented on Sources/FioriIntegrationCards/Components/Common/Model/Manifest.swift:
```diff
@@ -16,6 +17,8 @@ public class Manifest: Decodable, Identifiable, ObservableObject {
         return app.id
     }
     
+    public var baseURL: String?
```
`baseURL` is of type String?

MarcoEidinger commented on Sources/FioriIntegrationCards/Components/Common/Model/DataFetcher.swift:
```diff
@@ -40,8 +40,8 @@ public class DataFetcher: Decodable {
     
-    public func load() {
-        request?.send()
+    public func load(baseURL: String? = nil) {
```
`baseURL` is of type String?

MarcoEidinger commented on Sources/FioriIntegrationCards/Components/Common/Model/Card.swift:
```diff
@@ -65,6 +65,21 @@ public enum Card: Decodable, Identifiable, CustomStringConvertible {
         }
     }
+
+    public func loadDataIfNeeded(baseURL: String?) {
```
`baseURL` is of type String?

MarcoEidinger commented on Sources/FioriIntegrationCards/Components/Common/Model/BaseCard.swift:
```diff
@@ -138,7 +157,7 @@ open class BaseBaseCard: Decodable, ObservableObject, Identifiable {
     public let headerPublisher = CurrentValueSubject<CurrentValueSubject<(Data, String?)?, Never>?, Never>(nil)
     public let contentPublisher = CurrentValueSubject<CurrentValueSubject<(Data, String?)?, Never>?, Never>(nil)
-    
+    public let baseURL: CurrentValueSubject<String?, Never> = CurrentValueSubject<String?, Never>(nil)
```
`baseURL` is of type String?

MarcoEidinger commented on Sources/FioriIntegrationCards/Components/Common/Model/Manifest.swift:
```diff
@@ -29,6 +32,50 @@ public class Manifest: Decodable, Identifiable, ObservableObject {
+        }
+        self.app        = _model.app
+        self.card       = _model.card
+        self.baseURL    = path.absoluteString
```
converting URL to String

MarcoEidinger commented on Sources/FioriIntegrationCards/Components/Common/Model/Request.swift:
```diff
@@ -43,20 +43,60 @@ class Request: Decodable {
+        var urlRequest: URLRequest!
+        do {
+            if let _baseURL = baseURL {
+                urlRequest = try NetworkRouter.getURLRequest(with: self, baseURL: URL(string: _baseURL))
```
finally we load data with `baseURL` of type URL

MarcoEidinger commented on Sources/FioriIntegrationCards/Components/Common/Model/Request.swift:
```diff
@@ -43,20 +43,60 @@ class Request: Decodable {
+        }
+    }
+    
+    func getDataFromBundle(bundleURL: URL) -> Data? {
```
finally we load data with `bundleURL` of type URL

MarcoEidinger commented on Sources/FioriIntegrationCards/Utils/String+Extensions.swift:
```diff
@@ -9,6 +9,16 @@ import Foundation
 extension String {
     
+    var isValidURL: Bool {
+        let detector = try! NSDataDetector(types: NSTextCheckingResult.CheckingType.link.rawValue)
```
TODO: better guard condition?

MarcoEidinger commented:
@ShawnMa16 please remove the respective limitation note in README.md as your PR resolves it :)

> FioriIntegrationCards currently handles only data which is in-line json; must be augmented to support resolving relative data files, and remote URIs

ShawnMa16 commented:
> @ShawnMa16 please remove the respective limitation note in README.md as your PR resolves it :)
> 
> > FioriIntegrationCards currently handles only data which is in-line json; must be augmented to support resolving relative data files, and remote URIs

Hey Marco, I've addressed the comments, can you take a look?

MarcoEidinger commented:
Thanks @ShawnMa16 ,  looks good. This PR is good for merging. I noticed that in some cases errors are getting logged via `print` and are not propagated to the consumer. As this is already the case in other code for integration cards I created https://github.com/SAP/cloud-sdk-ios-fiori/issues/71 to address this in one shot.

MarcoEidinger commented:
Closes #33

MarcoEidinger commented:
I squashed commits to one and kept the best commit description :)
